lib/matplotlib/backends/backend_macosx.py

```diff
@@ -30,26 +30,37 @@ def __init__(self, figure):
         FigureCanvasBase.__init__(self, figure)
         width, height = self.get_width_height()
         _macosx.FigureCanvas.__init__(self, width, height)
+        self._draw_pending = False
+        self._is_drawing = False
 
     def set_cursor(self, cursor):
         # docstring inherited
         _macosx.set_cursor(cursor)
 
-    def _draw(self):
-        renderer = self.get_renderer(cleared=self.figure.stale)
-        if self.figure.stale:
-            self.figure.draw(renderer)
-        return renderer
-
     def draw(self):
-        # docstring inherited
-        self._draw()
-        self.flush_events()
+        """Render the figure and update the macosx canvas."""
+        # The renderer draw is done here; delaying causes problems with code
+        # that uses the result of the draw() to update plot elements.
+        if self._is_drawing:
+            return
+        with cbook._setattr_cm(self, _is_drawing=True):
+            super().draw()
+        self.update()
 
-    # draw_idle is provided by _macosx.FigureCanvas
+    def draw_idle(self):
+        # docstring inherited
+        if (getattr(self, '_draw_pending', False) or
+                getattr(self, '_is_drawing', False)):
+            return
+        self._draw_pending = True
+        with self._idle_draw_cntx():
+            self._draw_pending = False
+            self.draw()
```

Comment:
This should not call `draw` as `draw_idle` should always be "cheap" .

Reply:
I think I took this from the qt_backend, which appears to do the same `self.draw()` call from within `draw_idle`? I'm not entirely clear where the "pending" of draw_idle() should take place here, whether it should be within the Python context managers, or if perhaps I'm just missing putting this into the Mac-specific event loop similar to the singleShot call from QT?

https://github.com/matplotlib/matplotlib/blob/ec99c15612f8d8d5267ce4e5a9deb309fad4f5f4/lib/matplotlib/backends/backend_qt.py#L450-L461

Reply:
I think the logic with Qt is that every call to `canvas.draw_idle` set `self._draw_pending` to `True` and then schedules a single-shot timer to run `canvas._draw_idle`.  Thus many calls to `draw_idle` within a block of Python (which runs without letting the event loop spin) will stack up a bunch of single-shot timers.  When ever the main loop gets to run (which is the earliest we would be able to paint the screen anyway!) the timers start to expire.  The first one actually does the update and the rest short-circuit out.

I suspect that you can do the same single-shot trick with OSX?

Reply:
Thank you for the clear explanation! That makes sense to me (stacking a bunch of callbacks in the event loop, and those callbacks depend on whether the draw has already happened or not). Let me know if you think the implementation is wrong here, or if I misunderstood.